File: leetcode_alg/solution19.cpp
//
// Created by wangheng on 2021/6/20.
// 一趟遍历,删除链表倒数第n个节点 : 双指针方式解决
#include "iostream"
#include "listnode.h"

using namespace std;

class Solution {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        ListNode *node = new ListNode(-1, head);
        ListNode *p = node;
        ListNode *q = node;
        for (int i = 0; i < n + 1; i++) {
            q = q->next;
        }
        while (q != nullptr) {
            p = p->next;
            q = q->next;
        }
        ListNode *delNode = p->next;
        p->next = delNode->next;
        delete delNode;
        ListNode *realHead = node->next;
        delete node;
        return realHead;
    }
};

int main() {
    auto s = new Solution;
    ListNode *head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    ListNode *p = s->removeNthFromEnd(head, 5);
    while (p != nullptr) {
        cout << p->val << "->";
        p = p->next;
    }
    cout << endl;
    return 0;
}

File: golang/solution/solution19.go
package solution

import "fmt"

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	if head == nil {
		return head
	}
	if head.Next == nil && n == 1 {
		return nil
	}
	var newHead *ListNode = &ListNode{0, head}
	var slow, fast *ListNode = newHead, newHead
	for i := 0; i < n; i++ {
		fast = fast.Next
	}
	for fast.Next != nil {
		slow = slow.Next
		fast = fast.Next
	}
	slow.Next = slow.Next.Next
	return newHead.Next
}

func Solution19() {
	var list *ListNode = &ListNode{1, &ListNode{2, &ListNode{3, &ListNode{4, nil}}}}
	var res *ListNode = removeNthFromEnd(list, 3)

	for res != nil {
		fmt.Println(res.Val)
		res = res.Next
	}
}
